### PR TITLE
State rollback

### PR DESCRIPTION
**Pull request by antonlerner**

### CODE REVIEW

File: state/processor.go [hunk 1]
```diff
@@ -98,7 +153,6 @@ func (tp *TransactionProcessor) randomSort(transactions Transactions) Transactio
 	return transactions
```
Comment (y0sher) [inline state/processor.go:148]: `transactions[i], transactions[swp] = transactions[swp], transactions[i]` works in go :)

File: state/processor.go [hunk 2]
```diff
@@ -114,20 +168,42 @@ func (tp *TransactionProcessor) coalesceTransactionsBySender(transactions Transa
...
+			sortedOriginByTransactions = append(sortedOriginByTransactions, trans.Origin)
+			senderPut[trans.Origin] =true
+		}
+	}
+
+
+	for _, origin := range sortedOriginByTransactions {
+		for _, trns := range trnsBySender[origin] {
+			//todo: should we abort all transaction processing if we failed this one?
+			err := tp.ApplyTransaction(trns)
```
Comment (y0sher) [inline state/processor.go:183]: maybe lock the whole thing instead lock/unlock multiple times

File: state/processor.go [hunk 3]
```diff
@@ -49,27 +53,79 @@ type PseudoRandomizer interface {
...
+		rootHash: common.Hash{},
+		stateQueue: list.List{},
+		db : db.TrieDB(),
+		mu : sync.Mutex{},
 	}
 }
 
 //should receive sort predicate
-func (tp *TransactionProcessor) ApplyTransactions(transactions Transactions) error{
+func (tp *TransactionProcessor) ApplyTransactions(layer LayerID, transactions Transactions) error{
```
Comment (y0sher) [inline state/processor.go]: comment about this method relationship with `Reset` and why it's locked.

File: state/processor.go [hunk 4]
```diff
@@ -49,27 +53,79 @@ type PseudoRandomizer interface {
...
+		defer tp.mu.Unlock()
+		newState, err := New(state, tp.globalState.db)
+		log.Info("reverted, new root %x", newState.IntermediateRoot(false))
+		if err != nil {
+			panic("wtf")
+		}
+
+		tp.globalState = newState
+		//lock mutex
+		tp.pruneAfterRevert(layer)
```
Comment (y0sher) [inline state/processor.go:124]: `defer tp.mu.Unlock()` happens after this. is this by design ?

File: state/processor.go [hunk 5]
```diff
@@ -49,27 +53,79 @@ type PseudoRandomizer interface {
...
+}
 
+func (tp *TransactionProcessor) Reset(layer LayerID){
+	if state, ok := tp.prevStates[layer]; ok {
+		tp.mu.Lock()
+		defer tp.mu.Unlock()
+		newState, err := New(state, tp.globalState.db)
+		log.Info("reverted, new root %x", newState.IntermediateRoot(false))
+		if err != nil {
+			panic("wtf")
```
Comment (y0sher) [inline state/processor.go]: why not return an error ?

File: state/processor.go [hunk 6]
```diff
@@ -114,20 +168,42 @@ func (tp *TransactionProcessor) coalesceTransactionsBySender(transactions Transa
 	return trnsBySender
 }
 
-func (tp *TransactionProcessor) Process(transactions Transactions, trnsBySender map[common.Address][]*Transaction) error{
-	bySender := make(map[common.Hash]bool)
+func (tp *TransactionProcessor) Process(transactions Transactions, trnsBySender map[common.Address][]*Transaction) {
+	senderPut := make(map[common.Address]bool)
```
Comment (y0sher) [inline state/processor.go]: if the bool doesn't mean anything, `struct{}` is cheaper

File: state/processor.go [hunk 7]
```diff
@@ -114,20 +168,42 @@ func (tp *TransactionProcessor) coalesceTransactionsBySender(transactions Transa
...
+		}
+	}
+
+
+	for _, origin := range sortedOriginByTransactions {
+		for _, trns := range trnsBySender[origin] {
+			//todo: should we abort all transaction processing if we failed this one?
+			err := tp.ApplyTransaction(trns)
+			if err != nil {
+				log.Error("transaction aborted: %v", err)
```
Comment (y0sher) [inline state/processor.go:186]: maybe count successful processes and return the number ? so 0 might help to not move forward

**Review by y0sher — CHANGES_REQUESTED** [review]
Check the comments

File: state/processor.go (hunk 5 above)
Comment (antonlerner) [reply state/processor.go]: This is not something recoverable i think. anyway I'd like for this to be a panic for now, if we notice we get this one a lot during our testing we can think of why. but processing shouldn't continue after

**Review by y0sher — APPROVED** [review]
Good job, you just have to rebase on `develop`

**Pull request merged**